File: src/parser.py
from src.types import *

class LineParser:
    def __init__(self, tokens: list[Token], ic_line: int, src_line: int, vm = None) -> None:
        self.tokens = tokens
        self.i = -1
        self.ic_line = ic_line
        self.src_line = src_line
        self.cur = tokens[0] if tokens else None
        self.next = tokens[1] if len(tokens) > 1 else None
        self.vm = vm
        
        self.advance()

    def advance(self) -> Token:
        self.i += 1
        self.cur = self.tokens[self.i] if self.i < len(self.tokens) else None
        self.next = self.tokens[self.i + 1] if self.i + 1 < len(self.tokens) else None
        return self.cur

    def eat(self, *type: TokenType) -> Token:
        last = self.cur
        if self.cur and self.cur.type in type:
            self.advance()
            return last
        else:
            raise Exception(f"Expected token of type {type}, got {self.cur}")

    def parsePointerArithmetic(self) -> BinOp:
        """
        expects self.cur to be lparen. it passes beyond rparen.
        """
        self.eat(TokenType.LPAREN)

        left = self.cur
        self.eat(TokenType.REGISTER, TokenType.IMMEDIATE)

        uses_arithmetic = False
        while self.cur and self.cur.type in (TokenType.PLUS, TokenType.MINUS):
            uses_arithmetic = True
            op = self.cur
            self.advance()
            right = self.cur
            self.eat(TokenType.REGISTER, TokenType.IMMEDIATE)
            left = BinOp(left, op, right)

        if not uses_arithmetic:
            left = BinOp(left, Token(TokenType.PLUS, "+"), Token(TokenType.IMMEDIATE, "0"))

        self.eat(TokenType.RPAREN)
        return left

    def getValue(self) -> Token | BinOp:
        """
        expects self.cur to be on a value token. it a beyond the full value
        """
        if self.cur and self.cur.type in [TokenType.REGISTER, TokenType.IMMEDIATE]:
            return self.eat(TokenType.REGISTER, TokenType.IMMEDIATE)
        elif self.cur and self.cur.type == TokenType.LPAREN:
            return self.parsePointerArithmetic()
        else:
            raise Exception(f"Expected value token, got {self.cur}")

    def parse(self) -> list[Instruction]:
        self.instructions: list[Instruction] = []

        while self.cur is not None:
            if self.cur.type == TokenType.INST:
                inst = self.cur
                self.advance()
                if inst.value == "mov":
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    mov_inst = Instruction(InstructionType.MOV, {"dest": left, "src": right})
                    self.instructions.append(mov_inst)

                elif inst.value == "add":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    add_inst = Instruction(InstructionType.ADD, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(add_inst)
                
                elif inst.value == "sub":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    sub_inst = Instruction(InstructionType.SUB, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(sub_inst)

                elif inst.value == "mul":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    mul_inst = Instruction(InstructionType.MUL, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(mul_inst)

                elif inst.value == "idiv":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    idiv_inst = Instruction(InstructionType.IDIV, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(idiv_inst)

                elif inst.value == "fdiv":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    fdiv_inst = Instruction(InstructionType.FDIV, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(fdiv_inst)

                elif inst.value == "mod":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    fdiv_inst = Instruction(InstructionType.MOD, {"dest": dest, "left": left, "right": right})
                    self.instructions.append(fdiv_inst)

                elif inst.value == "abs":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    src = self.getValue()
                    abs_inst = Instruction(InstructionType.ABS, {"dest": dest, "src": src})
                    self.instructions.append(abs_inst)

                elif inst.value in ["jmp", "je", "jne", "jg", "jl", "jge", "jle"]:
                    label = self.eat(TokenType.SYMBOL)
                    jmp_inst = Instruction(InstructionType(inst.value), {"label": label})
                    self.instructions.append(jmp_inst)
               
                elif inst.value == "cmp":
                    left = self.getValue()
                    self.eat(TokenType.COMMA)
                    right = self.getValue()
                    cmp_inst = Instruction(InstructionType.CMP, {"left": left, "right": right})
                    self.instructions.append(cmp_inst)

                elif inst.value == "load":
                    dest = self.getValue()
                    self.eat(TokenType.COMMA)
                    src = self.parsePointerArithmetic()
                    load_inst = Instruction(InstructionType.LOAD, {"dest": dest, "src": src})
                    self.instructions.append(load_inst)
                
                elif inst.value == "store":
                    src = self.getValue()
                    self.eat(TokenType.COMMA)
                    dest = self.parsePointerArithmetic()
                    store_inst = Instruction(InstructionType.STORE, {"dest": dest, "src": src})
                    self.instructions.append(store_inst)

                elif inst.value == "print":
                    start = self.getValue()
                    self.eat(TokenType.COMMA)
                    end = self.getValue()
                    self.instructions.append(Instruction(InstructionType.PRINT, {"start": start, "end": end}))

                elif inst.value == "bp":
                    self.instructions.append(Instruction(InstructionType.BREAKPOINT, {}))

                elif inst.value == "call":
                    label = self.eat(TokenType.SYMBOL)
                    call_inst = Instruction(InstructionType.CALL, {"label": label})
                    self.instructions.append(call_inst)

                elif inst.value == "ret":
                    self.instructions.append(Instruction(InstructionType.RET, {}))

                elif inst.value == "screen":
                    rows = self.getValue()
                    self.eat(TokenType.COMMA)
                    cols = self.getValue()
                    self.eat(TokenType.COMMA)
                    memory_addr = self.getValue()
                    self.instructions.append(Instruction(InstructionType.SCREEN, {"rows": rows, "cols": cols, "memory_addr": memory_addr}))

                elif inst.value == "render":
                    self.instructions.append(Instruction(InstructionType.RENDER, {}))

                elif inst.value == "push":
                    src = self.getValue()
                    push_inst = Instruction(InstructionType.PUSH, {"src": src})
                    self.instructions.append(push_inst)

                elif inst.value == "pop":
                    dest = self.getValue()
                    pop_inst = Instruction(InstructionType.POP, {"dest": dest})
                    self.instructions.append(pop_inst)

                elif inst.value == "dec":
                    dest = self.getValue()
                    dec_inst = Instruction(InstructionType.DEC, {"dest": dest})
                    self.instructions.append(dec_inst)
                
                elif inst.value == "inc":
                    dest = self.getValue()
                    inc_inst = Instruction(InstructionType.INC, {"dest": dest})
                    self.instructions.append(inc_inst)

                elif inst.value == "input":
                    dest = self.getValue()
                    input_inst = Instruction(InstructionType.INPUT, {"dest": dest})
                    self.instructions.append(input_inst)

                elif inst.value == "time":
                    dest = self.getValue()
                    time_inst = Instruction(InstructionType.TIME, {"dest": dest})
                    self.instructions.append(time_inst)




            elif self.cur and self.cur.type == TokenType.SYMBOL and self.next and self.next.type == TokenType.COLON:
                label = self.cur.value
                self.vm.labels[label] = self.ic_line
                self.advance()
                self.advance()



            else:
                raise Exception(f"Unexpected token: {self.cur}")

        return self.instructions

File: c++/src/vm/parser.cpp
#ifndef PARSER_CPP
#define PARSER_CPP

#include "parser.hpp"

LineParser::LineParser(std::vector<Token> tokens)
{
    this->tokens = tokens;

    this->advance();
}


Token LineParser::advance()
{
    Token null_token;
    null_token.token_type = TokenType::null;
    this->i++;
    this->cur = (this->i < this->tokens.size()) ? this->tokens[this->i] : null_token;
    this->next = (this->i+1 < this->tokens.size()) ? this->tokens[this->i+1] : null_token;
    return this->cur;
}

/*
Checks if this->cur.token_type is equal to the proivded value list, and throws an exception if not. it stores this->cur and advances. returns the stores this->cur. 
*/
Token LineParser::eat(TokenType type)
{
    Token last = this->cur;
    if (this->cur.token_type != type)
    {
        std::cout << "Expected token of type " << tokenTypeRepr(type) << ", received " << tokenTypeRepr(this->cur.token_type) << std::endl;
        exit(0);
    }
    this->advance();
    return last;
}

/*
Checks if this->cur.token_type is in the provided list, and throws an exception if not. it stores this->cur and advances. returns the stores this->cur. 
*/
Token LineParser::eats(std::vector<TokenType> types)
{
    Token last = this->cur;
    if (! inTokenTypes(this->cur.token_type, types))
    {
        std::cout << "Expected token of type " << tokenTypesRepr(types) << ", received " << tokenTypeRepr(this->cur.token_type) << std::endl;
        exit(0);
    }
    this->advance();
    return last;
}

/*
expects cur to be LPAREN. passes beyond RPAREN
*/
Node LineParser::parsePointerArithmetic()
{
    this->eat(TokenType::LPAREN);
    
    Node left = this->cur;
    this->eats({TokenType::REGISTER, TokenType::IMMEDIATE});


    while (inTokenTypes(this->cur.token_type, {TokenType::PLUS, TokenType::MINUS}))
    {
        BinOp bin_op;
        bin_op.left = left;
        
        bin_op.op = this->cur;
        this->eats({TokenType::PLUS, TokenType::MINUS});
        
        bin_op.right = this->cur;
        this->eats({TokenType::REGISTER, TokenType::IMMEDIATE});
        
        left = std::make_shared<BinOp>(bin_op);
    }

    this->eat(TokenType::RPAREN);
    return left;    
}

/*
Expects to be on either a value token (i.e. register or imm) or on LPAREN. It passes beyond the full value, including if givin, RPAREN. 
*/
Node LineParser::getValue()
{
    if (inTokenTypes(this->cur.token_type, {TokenType::REGISTER, TokenType::IMMEDIATE}))
    {
        Token last = this->cur;
        this->advance();
        return last;
    }
    else if (this->cur.token_type == TokenType::LPAREN)
    {
        return this->parsePointerArithmetic();
    }
    else
    {
        std::cout << "Expected value token, received " << tokenTypeRepr(this->cur.token_type) << " " << tokenRepr(this->cur) << std::endl;
        exit(1);
    }
}

std::vector<InstructionIr> LineParser::parse()
{
    std::vector<InstructionIr> instructions = {};
    while (this->cur.token_type != TokenType::null)
    {
        InstructionIr inst;
        if (this->cur.value == "mov")
        {
            inst.type = InstructionType::MOV;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "add")
        {
            inst.type = InstructionType::ADD;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "sub")
        {
            inst.type = InstructionType::SUB;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "mul")
        {
            inst.type = InstructionType::MUL;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "idiv")
        {
            inst.type = InstructionType::IDIV;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "fdiv")
        {
            inst.type = InstructionType::FDIV;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "mod")
        {
            inst.type = InstructionType::MOD;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg3  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "abs")
        {
            inst.type = InstructionType::MUL;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jmp")  
        {
            inst.type = InstructionType::JMP;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "je")  
        {
            inst.type = InstructionType::JE;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jne")  
        {
            inst.type = InstructionType::JNE;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jg")  
        {
            inst.type = InstructionType::JG;
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jge")  
        {
            inst.type = InstructionType::JGE;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jl")  
        {
            inst.type = InstructionType::JL;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "jle")  
        {
            inst.type = InstructionType::JLE;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "cmp")
        {
            inst.type = InstructionType::CMP;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "load")
        {
            inst.type = InstructionType::LOAD;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "store")
        {
            inst.type = InstructionType::STORE;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "print")
        {
            inst.type = InstructionType::PRINT;
            this->advance();
            inst.arg1 = this->getValue();
            this->eat(TokenType::COMMA);
            inst.arg2  = this->getValue();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "bp")
        {
            inst.type = InstructionType::BREAKPOINT;
            this->advance();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "call")  
        {
            inst.type = InstructionType::CALL;
            this->advance();
            inst.arg1 = this->eat(TokenType::SYMBOL);
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "ret")  
        {
            inst.type = InstructionType::RET;
            this->advance();
            instructions.push_back(inst);
            continue;
        }
        else if (this->cur.value == "push")
        {
            inst.type = InstructionType::PUSH;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.value == "pop")
        {
            inst.type = InstructionType::POP;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.value == "dec")
        {
            inst.type = InstructionType::DEC;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.value == "inc")
        {
            inst.type = InstructionType::INC;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.value == "input")
        {
            inst.type = InstructionType::INPUT;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.value == "time")
        {
            inst.type = InstructionType::TIME;
            this->advance();
            inst.arg1 = this->getValue();
            instructions.push_back(inst);
        }
        else if (this->cur.token_type == TokenType::SYMBOL)
        {
            inst.type = InstructionType::LABEL;
            inst.arg1 = this->cur;
            this->advance();
            this->eat(TokenType::COLON);
            instructions.push_back(inst);
        }
        else
        {
            std::cout << "Unexpected token: " << tokenTypeRepr(this->cur.token_type) << " " << tokenRepr(this->cur) << std::endl;
            exit(0);
        }
    }
    return instructions;   
}

#endif